library/lua/test_util/expect.lua

```diff
@@ -0,0 +1,234 @@
+    local keys, known_eq = {}, {}
+    local matched, keys_at_diff, diff = table_eq_recurse(a, b, keys, known_eq)
+    if matched then return true end
+    local keystr = '['..keys_at_diff[1]..']'
+    for i=2,#keys_at_diff do keystr = keystr..'['..keys_at_diff[i]..']' end
+    return false, comment,
+            ('key %s: "%s" ~= "%s"'):format(keystr, diff[1], diff[2])
+end
+
+local function matches(obj, matcher)
+    if type(matcher) == 'string' then
+        return tostring(obj):match(matcher)
+    elseif type(matcher) == 'function' then
+        return matcher(obj)
+    end
+    return false
+end
+
+-- matches errors thrown from the specified function. the check passes if an
+-- error is thrown and the thrown error matches the specified matcher.
+--
+-- matcher can be:
+--  a string, interpreted as a lua pattern that matches the error message
+--  a function that takes an err object and returns a boolean (true means match)
+function expect.error_match(matcher, func, comment)
+    local ok, err = pcall(func)
+    if ok then
+        return false, comment, 'no error raised by function call'
+    end
+    if matches(err, matcher) then return true end
+    local matcher_str = ''
+    if type(matcher) == 'string' then
+        matcher_str = (': "%s"'):format(matcher)
+    end
+    return false, comment,
+            ('error "%s" did not satisfy matcher%s'):format(err, matcher_str)
+end
+
+function expect.error(func, comment)
+    return expect.error_match('.*', func, comment)
+end
+
+-- matches error messages output from dfhack.printerr() when the specified
+-- callback is run. the check passes if all printerr messages are matched by
+-- specified matchers and no matchers remain unmatched.
+--
+-- matcher can be:
+--  a string, interpreted as a lua pattern that matches a message
+--  a function that takes the string message and returns a boolean (true means
+--    match)
+--  a populated table that can be used to match multiple messages (explained
+--    in more detail below)
+--
+-- if matcher is a table, it can contain:
+--  a list of strings and/or functions which will be matched in order
+--  a map of strings and/or functions to positive integers, which will be
+--    matched (in any order) the number of times specified
+--
+-- when this function attempts to match a message, it will first try the next
+-- matcher in the list (that is, the next numeric key). if that matcher doesn't
+-- exist or doesn't match, it will try all string and function keys whose values
+-- are numeric and > 0.
+--
+-- if a mapped matcher is matched, it will have its value decremented by 1.
+function expect.printerr_match(matcher, func, comment)
+    local saved_printerr = dfhack.printerr
+    local messages = {}
+    dfhack.printerr = function(msg) table.insert(messages, msg) end
+    dfhack.with_finalize(
+        function() dfhack.printerr = saved_printerr end,
+        func)
+    if type(matcher) ~= 'table' then matcher = {matcher} end
+    while messages[1] do
```

Comment:
Nitpick: should this be `while #messages > 0`? It's technically possible to call `printerr(nil)` or `printerr(false)` (although nobody _should_ do that, but maybe something equivalent to `printerr()` could happen accidentally). This is further complicated by the fact that `ipairs()` just... silently stops iterating if it encounters a `nil` value (`#` does not do that, fortunately), so I suspect `unmatched_messages` would end up empty.

Maybe something like this could work to at least error out in weird cases?
```lua
while #messages > 0 do 
    local msg = messages[1]
    assert(msg, 'printerr() called with no message')
```

Reply:
I did not know that about '#' -- I thought that it stopped short at a `nil` like `ipairs` would. I'll get this handled. Thanks!

Reply:
ok, so `nil` can never be in `messages` because `table.insert(messages, nil)` is a noop, so just `while #messages > 0` should be sufficient. need to add a few unit tests tho.

Reply:
done